app/controllers/application_controller/wait_for_task.rb

```diff
@@ -11,67 +11,71 @@ def wait_for_task
     @edit = session[:edit]  # If in edit, need to preserve @edit object
     raise Forbidden, _('Invalid input for "wait_for_task".') unless params[:task_id]
 
-    @edit = session[:edit]  # If in edit, need to preserve @edit object
-    session[:async] ||= {}
-    session[:async][:interval] ||= 1000 # Default interval to 1 second
-    session[:async][:params] ||= {}
+    async_interval = params[:async_interval] || 1000 # Default interval to 1 second
 
-    if MiqTask.find(params[:task_id].to_i).state != "Finished" # Task not done --> retry
-      browser_refresh_task(params[:task_id])
+    task = MiqTask.find(params[:task_id].to_i)
+    if task.state != "Finished" # Task not done --> retry
+      browser_refresh_task(params[:task_id], async_interval)
     else # Task done
-      session[:async][:params].each { |k, v| @_params[k] = v } # Merge in the original params and
-      send(session.fetch_path(:async, :params, :action)) # call the orig. method
+      async_params = task.context_data[:async_params]
```

Comment:
Should we delete the async_params out of the task?  They weren't there in the first place and we stuffed them in.  My concern is that the subsequent handler of the task might take the context data and send it wholesale, for example, to some provider operation.

Reply:
From what I can tell, widgets and others store a bunch of data in `context_data`. It looks like the wild west.

The only provider I see that even mentions `context_data` is manageiq-providers-workflow that uses it to store the kubernetes id.

I don't see anything that uses `context_data` whole sale. e.g. `grep -r 'context_data[^\[.]'` is mostly assignments and nothing is accessed in a whole sale way.

I'm still open to removing it. Not sure if a glitch could cause issues here. We would delete it from the task when we view the task?

---

`BottleneckEvent` does allow the full context to come out - but I didn't look too much into that since it seems irrelevant.

Reply:
Sorry - I misunderstood where this line of code fit in the flow.  So, at this point the task is already completed, and it technically doesn't matter.  I feel like it should be removed from a purity standpoint (i.e. whatever we add temporarily we shouild delete), but I won't hold up merge.

> We would delete it from the task when we view the task?

No I was thinking just literally changing this line to 

```suggestion
      async_params = task.context_data.delete(:async_params)
```

Reply:
ok. implemented but without the `tap` dance

Reply:
I think I merged before this landed cause I don't see it in the code.